plugins/statistics.koplugin/main.lua

```diff
@@ -284,321 +604,264 @@ function ReaderStatistics:addToMainMenu(menu_items)
-                    date_text = os.date("%Y-%m-%d", n)
-                elseif ptype == "weekly" then
-                    date_text = T(_("%1 Week %2"), os.date("%Y "), os.date(" %W", n))
-                elseif ptype == "monthly" then
-                    date_text = monthTranslation[os.date("%B", n)] .. os.date(" %Y", n)
-                else
-                    date_text = os.date("%Y-%m-%d", n)
-                end  --if ptype
-                if not dates[date_text] then
-                    dates[date_text] = {
-                        -- first pages of day is set to average of all pages
-                        read = avg_time_per_page,
-                        date = n,
-                        count = 1
-                    }
-                else
-                    local entry = dates[date_text]
-                    diff = n - entry.date
-                    -- page_max_time
-                    if (diff <= page_max_time and diff > 0) then
-                        entry.read = entry.read + n - entry.date
-                    else
-                        --add average time if time > page_max_time
-                        entry.read = avg_time_per_page + entry.read
-                    end  --if diff
-                    if diff < 0 then
-                        entry.read = avg_time_per_page + entry.read
-                    end
-                    entry.date = n
-                    entry.count = entry.count + 1
-                end  --if not dates[]
-            end  -- for sorted_performance_in_pages
-        end  -- if book_status
-    end  --for pairs(ReadHistory.hist)
-    return dates
-end
-
-local function getDatesForBook(book)
-    local dates = {}
-    local sorted_performance_in_pages = {}
-    local diff
-    local read_pages = util.tableSize(book.performance_in_pages)
-    for k, v in pairs(book.performance_in_pages) do
-        table.insert(sorted_performance_in_pages, k)
-    end
-    -- sort table by time (unix timestamp)
-    table.sort(sorted_performance_in_pages)
-    for i, n in pairs(sorted_performance_in_pages) do
-        local date_text = os.date("%Y-%m-%d", n)
-        if not dates[date_text] then
-            dates[date_text] = {
-                -- first pages of day is set to average of all pages
-                read = book.total_time_in_sec / read_pages,
-                date = n,
-                count = 1
-            }
+    local one_day = 86400 -- one day in seconds
+    local period_begin = now_stamp - ((sdays-1) * one_day) - from_begin_day
+    local modifier
+    if ptype == "daily" or ptype == "daily_weekday" then
+        modifier = '%Y-%m-%d'
+    elseif ptype == "weekly" then
+        modifier = '%Y-%W'
+    elseif ptype == "monthly" then
+        modifier = '%Y %m'
+    end
+    self:insertDB(self.id_curr_book)
+    local conn = SQ3.open(db_location)
+    local result_book = conn:exec(string.format("SELECT strftime('%s', start_time, 'unixepoch', 'localtime') AS dates, " ..
```

Comment:
Again, anything more complex I think would be really improved by making it a multiline SQL query string. Note that my suggestions are just about the general principle of the thing and offer no concrete statement about the specifics of formatting other than that having some really improves clarity.

```lua
[[SELECT Strftime('%s', start_time, 'unixepoch', 'localtime') AS dates,
    Count(DISTINCT page)                                      AS pages,
    Sum(period)                                               AS periods,
    start_time
FROM   page_stat
WHERE  start_time >= '%s'
GROUP  BY dates
ORDER  BY dates DESC;]]
```